### PR TITLE
[WIP] nspawn: Add --system-call-filter option

### PR DESCRIPTION
**Pull request by nhlfr**
Just showing my idea and progress, not ready for submitting upstream yet.

---

The --system-call-filter option allows to override the default
seccomp filter with the list predefined by the user.

Fixes #5163

### CODE REVIEW

File: src/nspawn/nspawn-seccomp.c [hunk 1]
```diff
@@ -36,104 +36,162 @@
...
-                { CAP_SYS_RAWIO,  SCMP_SYS(iopl)                },
-                { CAP_SYS_RAWIO,  SCMP_SYS(pciconfig_iobase)    },
-                { CAP_SYS_RAWIO,  SCMP_SYS(pciconfig_read)      },
-                { CAP_SYS_RAWIO,  SCMP_SYS(pciconfig_write)     },
+const SyscallFilterSet default_syscall_filter = {
+        .name = "@nspawn_default",
+        .help = "System calls filter for systemd-nspawn",
+        .value =
+        "_sysctl\0"
+        "add_key"
```
Comment (alban) [inline src/nspawn/nspawn-seccomp.c]: What is the `\0`? I wonder why it's there for _sysctl but not for the other.

File: src/nspawn/nspawn-seccomp.h [hunk 2]
```diff
@@ -21,4 +21,23 @@
...
 
-int setup_seccomp(uint64_t cap_list_retain);
+#ifdef HAVE_SECCOMP
+#include "seccomp-util.h"
+
+typedef struct SyscallFilterSetCap {
+        SyscallFilterSet filter;
+        int capability;
+} SyscallFilterSetCap;
+#endif
```
Comment (alban) [inline src/nspawn/nspawn-seccomp.h]: Is there a reason to have the typedef SyscallFilterSetCap inside the "#ifdef"?

I am not quite sure about systemd, but in the Linux kernel header files, they try to have the definitions being defined both in #ifdef block and the #else block, sometimes with dummy struct/functions definitions for the #else block.

Comment (alban) [reply src/nspawn/nspawn-seccomp.h]: example of dummy function from http://www.linuxjournal.com/node/5780/print : 
```
#if ENABLE_BLAAH
#define blaah(x) if(x) printf("blaah %d\n",x)
#else
#define blaah(x) do { } while(0)
#endif

```

File: src/nspawn/nspawn-seccomp.c (hunk 1 above)
Comment (nhlfr) [reply src/nspawn/nspawn-seccomp.c]: My fault, I need to add `\0` everywhere, as it separates syscalls in filter, like here https://github.com/systemd/systemd/blob/master/src/shared/seccomp-util.c#L267-L647

File: src/nspawn/nspawn-seccomp.h (hunk 2 above)
Comment (nhlfr) [reply src/nspawn/nspawn-seccomp.h]: Well, I thought that seccomp-util.h can be included only when having seccomp headers, but that's not true, so I removed it. Thanks for pointing out.

**Pull request closed without merging**